subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c

```diff
@@ -538,3 +635,317 @@ int nrf_cloud_decode_data_endpoint(const struct nrf_cloud_data *input,
+			LOG_ERR("Invalid P-GPS array response format");
+			err = -EFTYPE;
+			goto cleanup;
+		}
+	} else if (get_string_from_obj(rsp_obj, PGPS_RCV_REST_HOST, &host_ptr) ||
+		   get_string_from_obj(rsp_obj, PGPS_RCV_REST_PATH, &path_ptr)) {
+		LOG_ERR("Invalid P-GPS REST response format");
+		err = -EFTYPE;
+		goto cleanup;
+	}
+
+	if (!host_ptr || !path_ptr) {
+		err = -ENOSTR;
+		goto cleanup;
+	}
+
+	if ((result->host_sz <= strlen(host_ptr)) ||
+	    (result->path_sz <= strlen(path_ptr))) {
+		err = -ENOBUFS;
+		goto cleanup;
+	}
+
+	strncpy(result->host, host_ptr, result->host_sz);
+	LOG_DBG("host: %s", log_strdup(result->host));
+
+	strncpy(result->path, path_ptr, result->path_sz);
+	LOG_DBG("path: %s", log_strdup(result->path));
+
+cleanup:
+	if (rsp_obj) {
+		cJSON_Delete(rsp_obj);
+	}
+	return err;
+}
+
+int get_string_from_array(const cJSON * const array, const int index,
+			  char **string_out)
+{
+	__ASSERT_NO_MSG(string_out != NULL);
+
+	cJSON *item = cJSON_GetArrayItem(array, index);
+
+	if (!cJSON_IsString(item)) {
+		return -EINVAL;
+	}
+
+	*string_out = item->valuestring;
+
+	return 0;
+}
+
+int get_string_from_obj(const cJSON * const obj, const char *const key,
+			char **string_out)
+{
+	__ASSERT_NO_MSG(string_out != NULL);
+
+	cJSON *item = cJSON_GetObjectItem(obj, key);
+
+	if (!cJSON_IsString(item)) {
+		return -EINVAL;
+	}
+
+	*string_out = item->valuestring;
+
+	return 0;
+}
+
+char * const nrf_cloud_format_cell_pos_req_payload(struct lte_lc_cells_info const * const inf,
+	size_t inf_cnt)
+{
+	if (!inf || !inf_cnt) {
+		return NULL;
+	}
+
+	char * payload = NULL;
+	cJSON *lte_array;
+	cJSON *payload_obj = cJSON_CreateObject();
+
+	lte_array = cJSON_AddArrayToObject(payload_obj, NRF_CLOUD_CELL_POS_JSON_KEY_LTE);
+	if (!lte_array) {
+		goto cleanup;
+	}
+
+	for (size_t i = 0; i < inf_cnt; i++) {
+		struct lte_lc_cell const * const cur = &(inf + i)->current_cell;
+		cJSON *lte_obj = cJSON_CreateObject();
+
+		if (!lte_obj) {
+			goto cleanup;
+		}
+		cJSON_AddItemToArray(lte_array, lte_obj);
+
+		/* required items */
+		if (json_add_num(lte_obj, NRF_CLOUD_CELL_POS_JSON_KEY_ECI, cur->id) ||
+		    json_add_num(lte_obj, NRF_CLOUD_CELL_POS_JSON_KEY_MCC, cur->mcc) ||
+		    json_add_num(lte_obj, NRF_CLOUD_CELL_POS_JSON_KEY_MNC, cur->mnc)) {
+			goto cleanup;
+		}
+
+		/* optional */
+		if (json_add_num(lte_obj, NRF_CLOUD_CELL_POS_JSON_KEY_TAC, cur->tac) ||
+		    json_add_num(lte_obj, NRF_CLOUD_CELL_POS_JSON_KEY_EARFCN, cur->earfcn) ||
+		    json_add_num(lte_obj, NRF_CLOUD_CELL_POS_JSON_KEY_RSRP, cur->rsrp) ||
+		    //json_add_num(lte_obj, NRF_CLOUD_CELL_POS_JSON_KEY_AGE, 0) ||
+		    //json_add_num(lte_obj, NRF_CLOUD_CELL_POS_JSON_KEY_GECI, cur->id) ||
+		    json_add_num(lte_obj, NRF_CLOUD_CELL_POS_JSON_KEY_RSRQ, cur->rsrq)) {
+			goto cleanup;
+		}
+
+		if ((cur->timing_advance != 65535) &&
+		    (json_add_num(lte_obj, NRF_CLOUD_CELL_POS_JSON_KEY_T_ADV,
+		     cur->timing_advance))) {
+			goto cleanup;
+		}
```

Comment:
I assume there will be some logic to add the reported neighbor cell information as well